lib/Containers/MerkleTree.cpp

```diff
@@ -910,36 +1146,95 @@ template <typename Hasher, std::uint64_t const BranchingBits>
 typename MerkleTree<Hasher, BranchingBits>::Meta&
 MerkleTree<Hasher, BranchingBits>::meta() noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(_buffer);
-  return *reinterpret_cast<Meta*>(_buffer.get());
+  return _data.meta;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
-typename MerkleTree<Hasher, BranchingBits>::Meta const&
-MerkleTree<Hasher, BranchingBits>::meta() const noexcept {
+MerkleTreeBase::Meta const& MerkleTree<Hasher, BranchingBits>::meta() const noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(_buffer);
-  return *reinterpret_cast<Meta const*>(_buffer.get());
+  return _data.meta;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
-typename MerkleTree<Hasher, BranchingBits>::Node&
-MerkleTree<Hasher, BranchingBits>::node(std::uint64_t index) noexcept {
+MerkleTreeBase::Node& MerkleTree<Hasher, BranchingBits>::node(std::uint64_t index) {
   // not thread-safe, lock buffer from outside
+  
+  // determine shard
+  std::uint64_t shard = shardForIndex(index);
+  TRI_ASSERT(shard < numberOfShards());
+
+  // lazily allocate backing memory
+  if (_data.shards.size() <= shard) {
+    // allocate a few shards at once
```

Comment:
The comment says that we allocate a few at once, but the code does not do this, unless the underlying container does it automatically. After 4 shards have been reached, this code just resizes one by one, as far as I see.

Reply:
Fixed.